provd/internal/services/gdm/gdm.go

```diff
@@ -0,0 +1,148 @@
+	pb.UnimplementedGdmServiceServer
+	conn           *dbus.Conn
+	privateConn    *dbus.Conn
+	displayManager dbus.BusObject
+	greeter        dbus.BusObject
+	userVerifier   dbus.BusObject
+}
+
+// New returns a new instance of the GDM service.
+func New(conn *dbus.Conn, opts ...Option) (*Service, error) {
+	s := &Service{conn: conn}
+
+	s.displayManager = s.conn.Object(consts.DbusGdmPrefix, dbus.ObjectPath("/org/gnome/DisplayManager/Manager"))
+
+	// Applying options, checking for errors in obtaining DBus objects
+	for _, opt := range opts {
+		if err := opt(s); err != nil {
+			return nil, err
+		}
+	}
+
+	// Ask GDM to open a session
+	call := s.displayManager.Call(consts.DbusGdmPrefix+".Manager.OpenSession", 0)
+	if call.Err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to call Manager: %s", call.Err)
+	}
+
+	busAddress, ok := call.Body[0].(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "failed to get bus address from Manager")
+	}
+
+	if busAddress == "foobar" {
+		return s, nil
+	}
+
+	// Establish a private connection to the session bus
+	var err error
+	s.privateConn, err = dbus.Dial(busAddress)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to connect to session bus: %s", err)
+	}
+	err = s.privateConn.Auth(nil)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to authenticate: %s", err)
+	}
+
+	// Ignoring the error here, since GDM's private bus doesn't support the standard "Hello" method. It is needed for the tests, though.
+	_ = s.privateConn.Hello()
+
+	// Ping the UserVerifier and Greeter objects
+	s.userVerifier = s.privateConn.Object(consts.DbusGdmPrefix+".UserVerifier", dbus.ObjectPath("/org/gnome/DisplayManager/Session"))
+	err = s.userVerifier.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping UserVerifier: %s", err)
+	}
+	s.greeter = s.privateConn.Object(consts.DbusGdmPrefix+".Greeter", dbus.ObjectPath("/org/gnome/DisplayManager/Session"))
+	err = s.greeter.Call(consts.DbusPeerPrefix+".Ping", 0).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to ping Greeter: %s", err)
+	}
+
+	return s, nil
+}
+
+// LaunchDesktopSession launches a desktop session for the given user and password.
+func (s *Service) LaunchDesktopSession(ctx context.Context, req *pb.GdmServiceRequest) (*emptypb.Empty, error) {
+	// Validate request
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "received a nil request")
+	}
+	if req.Username == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "received an empty username")
+	}
+
+	// Listen for signals on the session bus
+	signals := make(chan *dbus.Signal)
+	defer close(signals)
+	s.privateConn.Signal(signals)
+	defer s.privateConn.RemoveSignal(signals)
+
+	// Begin user verification
+	s.userVerifier.Call(consts.DbusGdmPrefix+".UserVerifier.BeginVerificationForUser", 0, "gdm-password", req.Username)
```

Comment:
@matthew-hagemann I suppose I should catch the error here too and add a test case as well?

Reply:
Ah, yes please 🙂